From 5e911381ecb77f317ee1e29ba1fe237400bed6ac Mon Sep 17 00:00:00 2001
From: Michael-CY Lee <michael-cy.lee@mediatek.com>
Date: Mon, 13 Jan 2025 14:11:36 +0800
Subject: [PATCH 057/123] mtk: mt76: mt7996: set single STA's TTLM on AP MLD

Add a driver operation, mt7996_set_sta_ttlm, to set a STA's TTLM on AP
MLD. Note that the AP MLD has handled the relationship between Adv-TTLM
and Neg-TTLM before calling the operation, so it can call
mt7996_mcu_peer_mld_ttlm_req() without any check.

Signed-off-by: Michael-CY Lee <michael-cy.lee@mediatek.com>
---
 mt7996/main.c | 14 ++++++++++++++
 1 file changed, 14 insertions(+)

diff --git a/mt7996/main.c b/mt7996/main.c
index 1a4e2904..c05e0672 100644
--- a/mt7996/main.c
+++ b/mt7996/main.c
@@ -2836,6 +2836,19 @@ mt7996_set_ttlm(struct ieee80211_hw *hw, struct ieee80211_vif *vif)
 	return ret;
 }
 
+static int
+mt7996_set_sta_ttlm(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
+		    struct ieee80211_sta *sta, struct ieee80211_neg_ttlm *neg_ttlm)
+{
+	struct mt7996_dev *dev = mt7996_hw_dev(hw);
+	int ret;
+
+	mutex_lock(&dev->mt76.mutex);
+	ret = mt7996_mcu_peer_mld_ttlm_req(dev, vif, sta, neg_ttlm);
+	mutex_unlock(&dev->mt76.mutex);
+	return ret;
+}
+
 static int
 mt7996_set_attlm(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
 		 u16 disabled_links, u16 switch_time, u32 duration)
@@ -3016,6 +3029,7 @@ const struct ieee80211_ops mt7996_ops = {
 	.change_sta_links = mt7996_mac_sta_change_links,
 	.set_qos_map = mt7996_set_qos_map,
 	.set_attlm = mt7996_set_attlm,
+	.set_sta_ttlm = mt7996_set_sta_ttlm,
 	.can_neg_ttlm = mt7996_can_neg_ttlm,
 	.set_ttlm = mt7996_set_ttlm,
 };
-- 
2.45.2

